lib/checks/lists/only-listitems.js

```diff
@@ -1,18 +1,84 @@
-var bad = [],
-	permitted = ['STYLE', 'META', 'LINK', 'MAP', 'AREA', 'SCRIPT', 'DATALIST', 'TEMPLATE'],
-	hasNonEmptyTextNode = false;
-
-virtualNode.children.forEach(({ actualNode }) => {
-	var nodeName = actualNode.nodeName.toUpperCase();
-	if (actualNode.nodeType === 1 && nodeName !== 'LI' && permitted.indexOf(nodeName) === -1) {
-		bad.push(actualNode);
-	} else if (actualNode.nodeType === 3 && actualNode.nodeValue.trim() !== '') {
-		hasNonEmptyTextNode = true;
-	}
-});
-
-if (bad.length) {
-	this.relatedNodes(bad);
+const ALLOWED_TAGS = [
+	'style',
+	'meta',
+	'link',
+	'map',
+	'area',
+	'script',
+	'datalist',
+	'template'
+];
+
+const getIsListItemRole = (r, t) => {
+	let out = false;
+	out = (r === 'listitem' || (t === 'li' && !r));
+	return out;
+};
+
+const getHasListItem = (hasListItem, tagName, isListItemRole) => {
+	return hasListItem || (tagName === 'li' && isListItemRole) || isListItemRole;
+}
+
+let base = {
+	badNodes: [],
+	isEmpty: true,
+	hasNonEmptyTextNode: false,
+	hasListItem: false,
+	liItemsWithRole: 0
+};
+
+let out = virtualNode.children
+	.reduce((out, { actualNode }) => {
+		/*eslint 
+		complexity: ["error", 11] 
+		*/
+		const tagName = actualNode.nodeName.toLowerCase();
+
+		switch (actualNode.nodeType) {
+		case 1:
+			if (!ALLOWED_TAGS.includes(tagName)) {
+				const role = (actualNode.getAttribute('role') || '').toLowerCase();
+				const isListItemRole = getIsListItemRole(role, tagName);
+				
+				out.hasListItem = getHasListItem(out.hasListItem, tagName, isListItemRole);
+
+				if (isListItemRole) {
+					out.isEmpty = false;
+				}
+				if (tagName === 'li' && !isListItemRole) {
+					out.liItemsWithRole++;
+				}
+				if (tagName !== 'li' && !isListItemRole) {
+					out.badNodes.push(actualNode);
+				}
+			}
+			break;
+		case 3:
+			if (actualNode.nodeValue.trim() !== '') {
+				out.hasNonEmptyTextNode = true;
+			}
+			break;
+		default:
+			break;
+		}
+		return out;
+	}, base);
+
+const virtualNodeChildrenOfTypeLi = virtualNode.children
+	.filter(({ actualNode }) => {
+		return actualNode.nodeName.toLowerCase() === 'li'
+	});
+
+const allLiItemsHaveRole = out.liItemsWithRole > 0 &&
+	virtualNodeChildrenOfTypeLi.length === out.liItemsWithRole;
+
+if (out.badNodes.length) {
+	this.relatedNodes(out.badNodes);
 }
 
-return !!bad.length || hasNonEmptyTextNode;
+return (
```

Comment:
It's really difficult to see if this code actually does what it should. Please break this up.